### DISCUSSION TITLE
Oqtane client side

### DISCUSSION THREAD

**Discussion by markagama** [oqtane/oqtane.framework#1715] [Q&A]
Hi, a few questions.
1.it appears that I have a lot of questions that might be FAQ. Is there a FAQ, I could not find one. If there is not, how can it be created and how can I contribute?
2. Before I found Oqtane, I was building various test environments using Visual Studio 2019, .Net 5 and Blazor template. The test environments were all blazor webassembly or single page application. I have noticed that Oqtane is quite different, is this because it was built with the server side blazor model before Microsoft had introduced the .Net 5 Blazor template? In Oqtane the wwwroot is under the server application - is this due to the original server side code development? Can it be moved so that Oqtane becomes similar to the Microsoft .Net 5 blazor template?
3. if Oqtane is server side how do I make it into a client side webassembly or SPA? I noticed in one document a statement that Oqtane could be run client side or server side, but I cannot find out how to make the change. To be more specific:
a. In appsettings.json change the Runtime from Server to WebAssembly
b. What about the RenderMode does it change or should it be deleted?
4. Should Oqtane be reworked now that Microsoft has made the new Blazor template available and clearly identified what a WebAssembly and SPA should look like?
5. Is it better to have two Oqtane code trees? One for WebAssembly / SPA and one for Server? Is Oqtane trying to be all things to everyone and trying to do too much?
6. I read that you have to manually add dependencies  to one of the client files when in WebAssembly mode with Oqtane. I noticed using the Microsoft Blazor template that it appears to do this automatically when in WebAssembly mode? Possibly this does not occur when adding nugets, but it appeared to do so.

One of the reasons why I ask, is that webassembly and SPA were created to speed up the applications, especially those that are heavy with client side controls and to ensure that users have an improved experience. I read some of the arguments here about server side but if server side is the future then why is webassembly/SPA even considered?

Has anyone carried out a comparative analysis between Drupal and DNN to identify why Drupal is so popular? It might be helpful to identify the key differences, key advantages and disadvantages. I remember people telling me to not consider DNN because it was visual basic, though I remember somewhere along the path it moved to c# and dotnet.

I'm certain that most of my questions are FAQs so I would be happy to add the answers to a FAQ.

**Comment by curlyfro** [comment]
lots your questions + more are answered in Shawn's youtube videos.  Search youtube for oqtane.  here's one:

[https://www.youtube.com/watch?v=gaRPnfSMQn8](https://www.youtube.com/watch?v=gaRPnfSMQn8)

**Reply by markagama** [reply]
thank you @curlyfro 
I'm hoping to contribute to a FAQ. I'm really interested to know if Oqtane should have two root trees, to more closely follow the code layout for WebAssembly, as I don't see the current code structure really being WebAssembly, it appears to be strongly server side focused with limited WebAssembly functionality if this is selected as the build mode.

**Comment by sbwalker** [comment]
@markagama comparing Oqtane to the default Blazor WebAssembly Visual Studio template app is like comparing apples to oranges - the default template is basically a "hello world" app with no useful functionality and Oqtane is a real-world application. Rather than answering your questions, I would encourage you to read some of the Blazor documentation starting with https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-5.0. If you read the WebAssembly section you will note that Oqtane is a  "hosted" Blazor WebAssembly app ( not a "standalone" Blazor WebAssembly app ). The reason it is hosted is because it requires the dynamic functionality of a *.cshtml on the server side to do the initial page render which is sent to the browser. The *.cshtml allows for the dynamic configuration of the hosting model, prerendering settings, and dynamic injection of static resources such as CSS and JavaScript into the initial page payload. To be clear, Oqtane is NOT a template or boilerplate application - it is a framework which contains extensibility points for integrating custom functionality.

**Comment by markagama** [comment]
Thank you for replying @sbwalker 

Whilst you state that the template creates a "hello world" app, it does provide a guide as to the thinking of the people that have created Blazor and it does clearly provide guidance on the distribution of logic. There is a significant difference between the template outcomes and Oqtane, which you do mention is a server side hosted model.

I'm not asking for defensive responses to my questions, I'm trying to understand what was done and why. You mention that Oqtane is based on the server side DNN design and this may not be a solution for those looking at WebAssembly and SPA.

One question that arises, which I'm sure myself and others will appreciate, is should there be two code trees, one for server side and one for standalone WebAssembly SPA?

Anyway, it would be really good to get a FAQ together, as it is very difficult to read about Oqtane at the moment.

**Reply by sbwalker** [reply]
@markagama I don't think I have ever said that Oqtane is based on the server-side DNN model - if I have could you please provide a link so that I am able to correct it? The Oqtane framework has concepts that are inspired by concepts in DNN - but the implementation is completely different. Oqtane is client/server - it has a distinct client project and a distinct server project - whereas DNN was 100% server-based. When running on WebAssembly the client project executes in the client browser and the server project executes on the server - so there are 2 separate code trees to satisfy the different workload requirements. However Oqtane is also architected to support running in Server mode as well - where the client project code actually executes on the server.

**Reply by markagama** [reply]
@sbwalker I think it is important to distinguish semantics so that others reading this in the future may understand what Oqtane is. You mention that Oqtane is a hosted model that utilises prerendering "on the server side to do the initial page render which is sent to the browser". The use of the phrase "on the server side" indicates it is a "server side hosted model".

It is important that people fully understand the implications of this, and what it means when looking for a Blazor WebAssembly / SPA framework.

I think my original point is why use Oqtane when it is a reworking of DNN to use Blazor. It appears to me that most developers will look and say, yup I'll stick with Drupal. If Oqtane was to take a step into the 21st century then it would have been great to see it become the first Blazor WebAssembly / SPA framework.

I'm moving ahead with Oqtane as it is the most advanced framework available that I can find using Blazor. At the same time, I'm going to create a Blazor templated WebAssembly / SPA app and work out a mechanism to be able to port my modules between the two environments in anticipation of a new framework appearing that is Blazor WebAssembly / SPA.

**Reply by sbwalker** [reply]
@markagama I apologize but I am having a very difficult time understanding your comments. I am very familiar with the various Blazor hosting models and whether a Blazor app is bootstrapped from a static html page or a dynamic cshtml page, that initial page is located on a server and the client browser needs to connect to it in order to get the initial page content - which includes, among other things, the reference to the Blazor JavaScript file for your preferred hosting model - whether that is Server or WebAssembly. So all I was trying to explain with my earlier comment is that Oqtane uses a cshtml page because there is a variety of dynamic functionality it offers on startup which would not be possible if used a static html page. When running Oqtane on WebAssembly it is fully leveraging the client-side model and is only interacting with the server to retrieve data - all UI rendering and events are happening natively in the browser. Oqtane is 100% architected as a SPA framework. It bears no resemblance to DNN from an architectural perspective whatsoever.

**Reply by markagama** [reply]
@sbwalker thx for the reply, it is clear now. As I mentioned it would be great if there was a FAQ, but I've mentioned this, and will move on.
